src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/OverloadMethodsDeclarationOrderCheck.java

```diff
@@ -117,36 +218,184 @@ public void visitToken(DetailAST ast) {
+                final IndexAndLine previousMeta = knownMethods.put(methodKey, indexAndLine);
+                if (previousMeta != null
+                    && currentIndex - previousMeta.getIndex() > allowedDistance) {
+                    final int previousLineWithOverloadMethod = previousMeta.getLineNo();
+                    log(currentToken, MSG_KEY, previousLineWithOverloadMethod);
                 }
-                methodIndexMap.put(methodName, currentIndex);
-                methodLineNumberMap.put(methodName, currentToken.getLineNo());
             }
             currentToken = currentToken.getNextSibling();
         }
     }
 
+    /**
+     * Compare the modifiers on a given method with the {@code modifierGroups} and find the
+     * first match to create a unique identifier for the group and method name combination.
+     *
+     * @param currentToken assumed to be a method
+     * @return string that uniquely represents the {@code modifierGroup} and method name
+     */
+    private String extractMethodDescription(DetailAST currentToken) {
+        final DetailAST method = currentToken.findFirstToken(TokenTypes.IDENT);
+        final String methodName = method.getText();
+
+        String descriptor = DEFAULT_MODIFIER_GROUP_PATTERN + SEPARATOR + methodName;
+        final String methodModifiers = extractModifiersString(currentToken);
+        for (Pattern pattern : modifierGroups) {
+            final Matcher matcher = pattern.matcher(methodModifiers);
+            if (matcher.find()) {
+                // method belongs to the first overload group which matches its modifiers
+                descriptor = pattern.pattern() + SEPARATOR + methodName;
+                break;
+            }
+        }
+        return descriptor;
+    }
+
+    /**
+     * Extract the modifier(s) of a method definition into a string.
+     *
+     * @param token expected to be TokenType.METHOD_DEF .
+     * @return modifier string for {@code token} with 'package' scope added if needed
+     */
+    private static String extractModifiersString(final DetailAST token) {
+        final List<String> collectedModifiers = getModifiers(token);
+        final boolean isInsideInterface =
+            token.getParent().getParent().getType() == TokenTypes.INTERFACE_DEF;
+
+        if (Collections.disjoint(SCOPE_MODIFIERS, collectedModifiers)) {
+            String defaultModifier = PACKAGE_MODIFIER;
+            if (isInsideInterface) {
+                defaultModifier = PUBLIC_MODIFIER;
+            }
+            collectedModifiers.add(0, defaultModifier);
+        }
+
+        // methods in interfaces can only have public, abstract, default, private,
+        // static modifiers. abstract is only valid when used with just the public
+        // modifier, so simply add it to the end of the list if it is missing
+        if (isInsideInterface && missingAbstractModifier(collectedModifiers)) {
```

Comment:
We check for `public` inside `missingAbstractModifier` https://github.com/checkstyle/checkstyle/pull/12086/files#diff-7fb7836b6edbcd2dfadb3ab0afa15939587026819fb95b06770c48f7dbb7e04eR353 . According to [JLS](https://docs.oracle.com/javase/specs/jls/se9/html/jls-9.html#jls-9.4), An interface method cannot be abstract if it is static, private, or has default. 

> An interface method lacking a private, default, or static modifier is implicitly abstract. Its body is represented by a semicolon, not a block. It is permitted, but discouraged as a matter of style, to redundantly specify the abstract modifier for such a method declaration

Reply:
we could simplify this to 
```java
if (ScopeUtil.isInInterfaceBlock(token) && List.of(Scope.PUBLIC.getName()).equals(collectedModifiers)) {
  collectedModifiers.add(ABSTRACT_MODIFIER);
}
```

Reply:
> We check for public inside missingAbstractModifier 

I see now, so the real issue is the bad method name. `missingAbstractModifier` would have been better to be named `hasOnlyPublicModifier`. You named it in the context of its only usage instead of what it does.

It doesn't seem to matter if this method is going away.

> we could simplify this to the below

With the usage of `ScopeUtil.getDefaultScope`, I would hope most of the custom code would go away.

Must `collectedModifiers` contain all modifiers, including non-explicit modifiers? How do we determine what order non-explicit modifiers are in? Are they always at the end, beginning, etc...? This has to be explicitly defined in the documentation with examples to show the user.
@romani ping

Reply:
I don't remember where I saw it, but I thought I saw somewhere that implicit modifiers should be added in the JLS recommended order

Reply:
Package makes sense, but the other implicit modifiers took me off guard. Your basically using similar detection defined in Scope, ClassMemberImpliedModifier, InterfaceMemberImpliedModifier, RedundantModifier. The implied check has a whole nest of things like static, abstract, etc.. which can all be implied. https://checkstyle.org/config_modifier.html#InterfaceMemberImpliedModifier_Properties

I am just confirming this is intended before we go further.

Reply:
>  How do we determine what order non-explicit modifiers are in?

after reread of this issue, I started to think that original ideas to make order of modifiers matter might be is not good. I might originally proposed it to keep it simple. But it looks like it does not simplify logic. So I am ok to switch from regexp to simply collection/set of modifiers. It will make this module more independent from ModifierOrder module, that is good.

Reply:
@austinarbor , yes.
hint: better to not use string values, and probably use some enum for this, but your will see that make more sense during implementation.

Reply:
My thoughts:
we need to focus on explicit modifiers, as users not always know implicit rules.
Method without modifiers will be separate group. Methods are usually in class or interface, so same rules of implicit will be and all of them will be in separate group.l, as even without modifiers Check still demands grouping.
Order of groups is not defined, this is good for us now.
User will need to unify modifiers to keep code clean.
If live show us good examples of why implicit matter, we can extend later on.

-----

Let's do only explicit modifiers consideration.

Reply:
With this approach how do you group together methods without any modifiers (i.e. package)?